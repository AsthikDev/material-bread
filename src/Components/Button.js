import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Text, View, StyleSheet, ActivityIndicator } from 'react-native';
import withTheme from '../Theme/withTheme';

import Ripple from '../Abstract/Ripple';

class Button extends Component {
  static propTypes = {
    disabled: PropTypes.bool,
    disableRipple: PropTypes.bool,
    loading: PropTypes.bool,
    onPress: PropTypes.func,
    textStyle: PropTypes.object,
    textColor: PropTypes.string,
    theme: PropTypes.object,
    style: PropTypes.object,
    children: PropTypes.node,
    fullWidth: PropTypes.bool,
    compact: PropTypes.bool,
  };
  render() {
    const {
      disabled,
      disableRipple,
      loading,
      children,
      onPress,
      style,
      theme,
      textStyle,
      textColor,
      fullWidth,
      compact,
    } = this.props;

    let buttonTextColor = textColor ? textColor : theme.button.color;
    buttonTextColor = disabled ? 'rgba(0, 0, 0, 0.26)' : buttonTextColor;

    const rippleColor = textColor ? textColor : theme.button.color;
    const styles = StyleSheet.create({
      button: {
        ...theme.button,
        flex: 1,
        alignSelf: 'center',
        alignItems: 'center',
        justifyContent: 'center',
        width: fullWidth ? '100%' : 'auto',
        minWidth: compact ? 'auto' : 64,
        ...style,
      },
      text: {
        ...theme.buttonText,
        color: buttonTextColor,
        ...textStyle,
      },
      icon: {
        width: 16,
        flexDirection: 'row',
        alignItems: 'center',
        alignSelf: 'center',
        justifyContent: 'center',
        textAlign: 'center',
      },
    });

    return (
      <Ripple
        onPress={onPress}
        disabled={disabled || disableRipple || loading}
        rippleColor={rippleColor}
        rippleContainerBorderRadius={theme.button.borderRadius}
        style={styles.button}>
        {loading ? (
          <ActivityIndicator
            size="small"
            color={textColor}
            style={styles.icon}
          />
        ) : (
          <Text numberOfLines={1} style={styles.text}>
            {React.Children.map(children, child =>
              typeof child === 'string' ? child.toUpperCase() : child,
            )}
          </Text>
        )}
      </Ripple>
    );
  }
}

export default withTheme(Button);
