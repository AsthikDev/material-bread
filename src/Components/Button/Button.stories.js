import React from 'react';
import { View, Text } from 'react-native';
import { storiesOf } from '../../storybook/helpers/storiesOf';

import { Button, Icon } from '../../';
import Header from '../../storybook/components/Header';
import Container from '../../storybook/components/Container';

export default storiesOf('Components|Buttons', module)
  .addParameters({ jest: ['Button'] })

  .add('Outlined Button', () => (
    <Container>
      <Header title={'Outlined Button'} />

      <View
        style={{
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'space-between',
          flexWrap: 'wrap',
        }}>
        <Button text={'Outlined'} type="outlined" />
        <Button
          text={'Icon'}
          type="outlined"
          textColor={'#E91E63'}
          borderSize={2}
          icon={<Icon name="favorite" />}
        />
        <Button
          text={'Radius'}
          type="outlined"
          textColor={'#FF5722'}
          radius={20}
        />
        <Button text={'Dense'} type="outlined" textColor={'#673AB7'} dense />
        <Button
          text={'Loading'}
          type="outlined"
          textColor={'#009688'}
          loading
        />
        <Button text={'Disabled'} type="outlined" disabled />
      </View>
      <Button
        fullWidth
        text={'Outlined'}
        type="outlined"
        textColor={'#673AB7'}
        style={{ marginBottom: 8 }}
      />
    </Container>
  ))
  .add('Contained Button', () => (
    <Container>
      <Header title={'Contained Button'} />
      <View
        style={{
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'space-between',
          flexWrap: 'wrap',
        }}>
        <Button text={'contained'} type="contained" />
        <Button
          text={'Icon'}
          type="contained"
          color={'#E91E63'}
          icon={<Icon name="favorite" />}
        />
        <Button
          text={'Radius'}
          type="contained"
          color={'#FF5722'}
          radius={20}
        />
        <Button text={'Dense'} type="contained" color={'#673AB7'} dense />
        <Button text={'Loading'} type="contained" color={'#009688'} loading />
        <Button text={'Disabled'} type="contained" disabled />
      </View>
      <Button
        fullWidth
        text={'Contained'}
        type="contained"
        containerStyle={{ marginBottom: 8 }}
      />
    </Container>
  ))
  .add('custom', () => (
    <Container>
      <Header title={'Custom Button'} />
      <View
        style={{
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'space-between',
          flexWrap: 'wrap',
        }}>
        <Button
          style={{
            height: 100,
            width: 100,
            flexDirection: 'column',
            marginRight: 24,
          }}
          type={'outlined'}
          borderSize={4}
          radius={10}>
          <Icon name="cloud-upload" size={34} />
          <Text style={{ fontWeight: '600', textAlign: 'center' }}>
            Upload files
          </Text>
        </Button>
      </View>
    </Container>
  ));
