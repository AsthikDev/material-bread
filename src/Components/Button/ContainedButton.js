import React, { Component } from 'react';
import { Animated, Platform } from 'react-native';

import PropTypes from 'prop-types';
import withTheme from '../../Theme/withTheme';

import ButtonBase from './ButtonBase';
import Hoverable from '../../Abstract/Hoverable';
import color from 'color';

class ContainedButton extends Component {
  static propTypes = {
    disabled: PropTypes.bool,
    color: PropTypes.string,
    textColor: PropTypes.string,
    rippleColor: PropTypes.string,
    theme: PropTypes.object,
    onPressIn: PropTypes.func,
    onPressOut: PropTypes.func,
    containerStyle: PropTypes.styles,
    radius: PropTypes.number,
    fullWidth: PropTypes.bool,
  };
  state = {
    stateBackgroundColor: null,
    elevation: new Animated.Value(6),
    animatedShadowRadius: new Animated.Value(4.65),
    animatedShadowOpacity: new Animated.Value(0.27),
    animatedShadowHeight: new Animated.Value(3),
    animatedShadowWidth: new Animated.Value(1),
  };

  animateShadow(active) {
    const animationDuration = active ? 400 : 150;
    if (Platform.OS === 'android') {
      const elevation = active ? 10 : 6;
      Animated.timing(this.state.elevation, {
        toValue: elevation,
        duration: animationDuration,
      });
    } else {
      const {
        animatedShadowRadius,
        animatedShadowOpacity,
        animatedShadowHeight,
        animatedShadowWidth,
      } = this.state;

      const shadowRadius = active ? 6.65 : 4.65;
      const shadowShadowOpacity = active ? 0.27 : 0.35;
      const shadowHeight = active ? 5 : 3;
      const shadowWidth = active ? 1.5 : 1;

      Animated.parallel([
        Animated.timing(animatedShadowRadius, {
          toValue: shadowRadius,
          duration: animationDuration,
        }),
        Animated.timing(animatedShadowOpacity, {
          toValue: shadowShadowOpacity,
          duration: animationDuration,
        }),
        Animated.timing(animatedShadowHeight, {
          toValue: shadowHeight,
          duration: animationDuration,
        }),
        Animated.timing(animatedShadowWidth, {
          toValue: shadowWidth,
          duration: animationDuration,
        }),
      ]).start();
    }
  }

  getShadowStyle() {
    const { theme } = this.props;
    const {
      animatedShadowRadius,
      animatedShadowOpacity,
      animatedShadowHeight,
      animatedShadowWidth,
      elevation,
    } = this.state;

    const style = {
      shadowColor: '#000',
      shadowOffset: {
        width: animatedShadowWidth,
        height: animatedShadowHeight,
      },
      shadowOpacity: animatedShadowOpacity,
      shadowRadius: animatedShadowRadius,
      elevation: elevation,
      backgroundColor: 'transparent',
      borderRadius: theme.button.borderRadius,
    };
    return style;
  }

  getButtonStyles() {
    const { theme } = this.props;

    const buttonStyles = {
      ...theme.containedButton,
      backgroundColor: this.getBackgroundColor(),
    };

    return buttonStyles;
  }

  getBackgroundColor() {
    const { color: userColor, disabled, theme } = this.props;
    const { stateBackgroundColor } = this.state;

    let backgroundColor = userColor ? userColor : theme.base.primary;

    backgroundColor = disabled ? 'rgba(0, 0, 0, 0.12)' : backgroundColor;

    return stateBackgroundColor ? stateBackgroundColor : backgroundColor;
  }

  getRippleColor() {
    const { rippleColor } = this.props;

    let implementedRippleColor = 'rgba(255, 255,255, 0.56)';

    return rippleColor ? rippleColor : implementedRippleColor;
  }

  getTextColor() {
    const { textColor, disabled } = this.props;

    let implementedTextColor = disabled ? 'rgba(0, 0, 0, 0.26)' : 'white';
    if (textColor) implementedTextColor = textColor;

    return implementedTextColor;
  }

  handleHover(toggle) {
    let implementedColor = toggle
      ? color(this.getBackgroundColor())
          .darken(0.15)
          .rgb()
          .string()
      : null;

    this.setState({ stateBackgroundColor: implementedColor });
  }

  handleOnPressIn = (...args) => {
    const { onPressIn } = this.props;
    this.animateShadow(true);

    if (onPressIn) onPressIn(args);
  };
  handleOnPressOut = (...args) => {
    const { onPressOut } = this.props;

    this.animateShadow(false);

    if (onPressOut) onPressOut(args);
  };

  render() {
    const {
      disabled,
      radius,
      theme,
      containerStyle,
      fullWidth,
      ...props
    } = this.props;

    return (
      <Hoverable
        onHoverIn={() => this.handleHover(true)}
        onHoverOut={() => this.handleHover(false)}>
        {() => (
          <Animated.View
            style={[
              disabled ? {} : this.getShadowStyle(),
              {
                alignSelf: fullWidth ? 'unset' : 'flex-start',
                borderRadius: radius ? radius : theme.button.borderRadius,
              },
              containerStyle,
            ]}>
            <ButtonBase
              typeRippleColor={this.getRippleColor()}
              typeTextColor={this.getTextColor()}
              typeButtonStyles={this.getButtonStyles()}
              onPressIn={this.handleOnPressIn}
              onPressOut={this.handleOnPressOut}
              disabled={disabled}
              radius={radius}
              theme={theme}
              fullWidth={fullWidth}
              {...props}
            />
          </Animated.View>
        )}
      </Hoverable>
    );
  }
}

export default withTheme(ContainedButton);
