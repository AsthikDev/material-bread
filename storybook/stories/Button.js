import React from 'react';
import { View } from 'react-native';
import { storiesOf } from '../helpers/storiesOf';

import { Button, Icon } from '../../src/index';
import Header from '../components/Header';
import Container from '../components/Container';

storiesOf('Buttons', module)
  .add('Text Button', () => (
    <Container>
      <Header title={'Text Button'} />

      <View
        style={{
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'space-between',
          flexWrap: 'wrap',
        }}>
        <Button text={'Text'} />
        <Button
          text={'Icon'}
          textColor={'#E91E63'}
          borderSize={2}
          icon={<Icon name="favorite" />}
        />
        <Button text={'Radius'} textColor={'#FF5722'} radius={20} />
        <Button text={'Dense'} ttextColor={'#673AB7'} dense />
        <Button text={'Loading'} textColor={'#009688'} loading />
        <Button text={'Disabled'} disabled />
      </View>
      <Button
        fullWidth
        text={'Text'}
        textColor={'#009688'}
        style={{ marginBottom: 8 }}
      />
    </Container>
  ))
  .add('Flat Button', () => (
    <Container>
      <Header title={'Flat Button'} />

      <View
        style={{
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'space-between',
          flexWrap: 'wrap',
        }}>
        <Button text={'Flat'} type="flat" />
        <Button
          text={'Icon'}
          type="flat"
          color={'#E91E63'}
          icon={<Icon name="favorite" />}
        />
        <Button text={'Radius'} type="flat" color={'#FF5722'} radius={20} />
        <Button text={'Dense'} type="flat" color={'#673AB7'} dense />
        <Button text={'Loading'} type="flat" color={'#009688'} loading />
        <Button text={'Disabled'} type="flat" disabled />
      </View>
      <Button
        fullWidth
        text={'Flat'}
        type="flat"
        color={'#E91E63'}
        style={{ marginBottom: 8 }}
      />
    </Container>
  ))
  .add('Outlined Button', () => (
    <Container>
      <Header title={'Outlined Button'} />

      <View
        style={{
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'space-between',
          flexWrap: 'wrap',
        }}>
        <Button text={'Outlined'} type="outlined" />
        <Button
          text={'Icon'}
          type="outlined"
          textColor={'#E91E63'}
          borderSize={2}
          icon={<Icon name="favorite" />}
        />
        <Button
          text={'Radius'}
          type="outlined"
          textColor={'#FF5722'}
          radius={20}
        />
        <Button text={'Dense'} type="outlined" textColor={'#673AB7'} dense />
        <Button
          text={'Loading'}
          type="outlined"
          textColor={'#009688'}
          loading
        />
        <Button text={'Disabled'} type="outlined" disabled />
      </View>
      <Button
        fullWidth
        text={'Outlined'}
        type="outlined"
        textColor={'#673AB7'}
        style={{ marginBottom: 8 }}
      />
    </Container>
  ))
  .add('Contained Button', () => (
    <Container>
      <Header title={'Contained Button'} />
      <View
        style={{
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'space-between',
          flexWrap: 'wrap',
        }}>
        <Button text={'contained'} type="contained" />
        <Button
          text={'Icon'}
          type="contained"
          color={'#E91E63'}
          icon={<Icon name="favorite" />}
        />
        <Button
          text={'Radius'}
          type="contained"
          color={'#FF5722'}
          radius={20}
        />
        <Button text={'Dense'} type="contained" color={'#673AB7'} dense />
        <Button text={'Loading'} type="contained" color={'#009688'} loading />
        <Button text={'Disabled'} type="contained" disabled />
      </View>
      <Button
        fullWidth
        text={'Contained'}
        type="contained"
        containerStyle={{ marginBottom: 8 }}
      />
    </Container>
  ));
