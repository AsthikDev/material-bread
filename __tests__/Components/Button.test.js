import React from 'react';
import Button from '../../src/Components/Button';

import renderer from 'react-test-renderer';

test('it renders Button', () => {
  const tree = renderer.create(<Button>Click Me</Button>).toJSON();
  expect(tree).toMatchSnapshot();
});
