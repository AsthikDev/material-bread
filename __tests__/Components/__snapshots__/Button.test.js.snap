// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders Button 1`] = `
<View
  accessible={true}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onRippleAnimation={[Function]}
  onStartShouldSetResponder={[Function]}
  pointerEvents="box-only"
  rippleCentered={false}
  rippleColor="#2196f3"
  rippleDuration={400}
  rippleFades={true}
  rippleOpacity={0.3}
  rippleSequential={false}
  rippleSize={0}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "transparent",
      "borderRadius": 4,
      "color": "#2196f3",
      "flex": 1,
      "height": 36,
      "justifyContent": "center",
      "minHeight": 36,
      "minWidth": 64,
      "padding": 8,
      "width": "auto",
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Object {
        "color": "#2196f3",
        "fontSize": 14,
        "fontWeight": "500",
        "letterSpacing": 1.25,
      }
    }
  >
    CLICK ME
  </Text>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "bottom": 0,
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "borderRadius": 4,
        },
      ]
    }
  />
</View>
`;

exports[`it renders Button fullwidth 1`] = `
<View
  accessible={true}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onRippleAnimation={[Function]}
  onStartShouldSetResponder={[Function]}
  pointerEvents="box-only"
  rippleCentered={false}
  rippleColor="#2196f3"
  rippleDuration={400}
  rippleFades={true}
  rippleOpacity={0.3}
  rippleSequential={false}
  rippleSize={0}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "transparent",
      "borderRadius": 4,
      "color": "#2196f3",
      "flex": 1,
      "height": 36,
      "justifyContent": "center",
      "minHeight": 36,
      "minWidth": 64,
      "padding": 8,
      "width": "auto",
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Object {
        "color": "#2196f3",
        "fontSize": 14,
        "fontWeight": "500",
        "letterSpacing": 1.25,
      }
    }
  >
    CLICK ME
  </Text>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "bottom": 0,
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "borderRadius": 4,
        },
      ]
    }
  />
</View>
`;

exports[`it renders loading Button 1`] = `
<View
  accessible={true}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onRippleAnimation={[Function]}
  onStartShouldSetResponder={[Function]}
  pointerEvents="box-only"
  rippleCentered={false}
  rippleColor="#2196f3"
  rippleDuration={400}
  rippleFades={true}
  rippleOpacity={0.3}
  rippleSequential={false}
  rippleSize={0}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "transparent",
      "borderRadius": 4,
      "color": "#2196f3",
      "flex": 1,
      "height": 36,
      "justifyContent": "center",
      "minHeight": 36,
      "minWidth": 64,
      "padding": 8,
      "width": "auto",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#2196f3"
    hidesWhenStopped={true}
    size="small"
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "textAlign": "center",
        "width": 16,
      }
    }
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "bottom": 0,
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "borderRadius": 4,
        },
      ]
    }
  />
</View>
`;

exports[`it renders outlined Button 1`] = `
<View
  accessible={true}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onRippleAnimation={[Function]}
  onStartShouldSetResponder={[Function]}
  pointerEvents="box-only"
  rippleCentered={false}
  rippleColor="#2196f3"
  rippleDuration={400}
  rippleFades={true}
  rippleOpacity={0.3}
  rippleSequential={false}
  rippleSize={0}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "transparent",
      "borderColor": "rgba(0,0,0,.29)",
      "borderRadius": 4,
      "borderWidth": 0.5,
      "color": "#2196f3",
      "flex": 1,
      "height": 36,
      "justifyContent": "center",
      "minHeight": 36,
      "minWidth": 64,
      "padding": 8,
      "paddingHorizontal": 16,
      "width": "auto",
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Object {
        "color": "#2196f3",
        "fontSize": 14,
        "fontWeight": "500",
        "letterSpacing": 1.25,
      }
    }
  >
    CLICK ME
  </Text>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "bottom": 0,
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "borderRadius": 4,
        },
      ]
    }
  />
</View>
`;
